### PR TITLE
ARROW-7303: [C++] Refactor CSV benchmarks to use Result APIs

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

**Comment by kszucs** [comment]
Shall we have a nightly GHA action where we exercise the benchmarks?

**Comment by wesm** [comment]
Not a bad idea. We don't care about the results in GHA but it would be good to know that they work

File: cpp/src/arrow/csv/converter_benchmark.cc [hunk 1]
```diff
@@ -79,8 +79,7 @@ static void BenchmarkConversion(benchmark::State& state,  // NOLINT non-const re
   std::shared_ptr<Converter> converter = *Converter::Make(type, options);
 
   while (state.KeepRunning()) {
-    std::shared_ptr<Array> result;
-    ABORT_NOT_OK(converter->Convert(parser, 0 /* col_index */, &result));
+    ASSIGN_OR_ABORT(auto result, converter->Convert(parser, 0 /* col_index */));
```
Comment (pitrou) [inline cpp/src/arrow/csv/converter_benchmark.cc]: Why not simply `auto result = *converter->Convert(parser, 0 /* col_index */);`?

Comment (bkietz) [reply cpp/src/arrow/csv/converter_benchmark.cc]: It seemed better to be explicit than let someone hunt on a `ValueOrDie` termination

Comment (pitrou) [reply cpp/src/arrow/csv/converter_benchmark.cc]: Don't you get the same kind of message in both cases?

Comment (bkietz) [reply cpp/src/arrow/csv/converter_benchmark.cc]: I think so. I'll just revert it

**Review by pitrou — APPROVED** [review]

**Comment by kszucs** [comment]
Created a tracking issue before we forget it https://issues.apache.org/jira/browse/ARROW-7337

**Pull request closed without merging**